### DISCUSSION TITLE
Sampling the unit sphere using square_to_uniform_sphere

### DISCUSSION THREAD

**Discussion by tomas16** [mitsuba-renderer/mitsuba3#968] [Q&A] [answered]
Hi, I want to uniformly sample the unit sphere. To this end, I wrote this function:

```python
def uniform_sphere_directions(n2):
    n1 = int(np.ceil(np.sqrt(n2)))
    offset = 1 / (2 * n1)
    x, y = dr.meshgrid(
        dr.linspace(mi.Float32, 0, 1, n1, endpoint=False) + offset,
        dr.linspace(mi.Float32, 0, 1, n1, endpoint=False) + offset,
    )
    pt = mi.Point2f(x, y)
    return mi.warp.square_to_uniform_sphere(pt)
```

When I call it with n2=36, I get this result:
![Screenshot 2023-11-07 at 2 32 12 PM](https://github.com/mitsuba-renderer/mitsuba3/assets/8531909/d6d63231-ca27-4f35-b95f-e7c7c4082af7)

I'm a bit surprised that spacing between points doesn't seem to be equal in the azimuthal and elevation directions. If one considers a full 2pi circle, then for horizontal cross-sections there are 6 points on that circle, whereas for vertical cross-sections, there are 12. If we're uniformly sampling the unit sphere, I wouldn't expect such asymmetry.

If I modify my function so the 2nd argument to meshgrid becomes `dr.linspace(mi.Float32, 0, 1, n1 // 2, endpoint=False) + offset * 2`, or in other words, the meshgrid becomes 6x3 instead of 6x6 before, then I get this:
![Screenshot 2023-11-07 at 2 36 26 PM](https://github.com/mitsuba-renderer/mitsuba3/assets/8531909/c4889ff5-d14a-4ac5-b35b-c96f55ef3b98)

This **looks** more uniform, but I'm puzzled as there doesn't seem to be a bug here. I went and looked at the implementation in mitsuba:
```cpp
/// Uniformly sample a vector on the unit sphere with respect to solid angles
template <typename Value>
MI_INLINE Vector<Value, 3> square_to_uniform_sphere(const Point<Value, 2> &sample) {
    Value z = dr::fnmadd(2.f, sample.y(), 1.f),
          r = circ(z);
    auto [s, c] = dr::sincos(2.f * dr::Pi<Value> * sample.x());
    return { r * c, r * s, z };
}
```

This matches with the theory and implementation from the pbrt book: `SampleUniformSphere `, explained at https://pbr-book.org/4ed/Sampling_Algorithms/Sampling_Multidimensional_Functions#UniformlySamplingHemispheresandSpheres

So I guess my question is, what am I missing here, and does `square_to_uniform_sphere` really convert uniform samples on the unit square to uniform samples on the sphere?
Perhaps the mapping really is uniform, but there's a different density of the mapping for X and Y.
Any thoughts / insights would be greatly appreciated.

**Comment by ziyi-zhang** [comment, marked as answer]
Hi @tomas16 , 
The uniform warp guarantees that area patches in the sample space will be mapped to equal-area patches on the sphere -- on a sphere $dS = r^2 ~\sin{\theta} d\theta d\phi$, so along the polar axis points are denser near the equator. 
But this does not really imply $||dp/d\theta||=||dp/d\phi||$ -- the visual uniformity after mapping meshgrid points. In fact, the azimuthal angle covers a $2\pi$ range and the polar angle only covers a range of $\pi$, so the observed pattern is normal.

**Reply by tomas16** [reply]
Thanks!

So essentially if you randomly and uniformly sample the unit square, that translates into uniform sampling on the sphere.
However that's not the same as saying an equally spaced grid on the square translates into an equally spaced grid on the sphere.

In this case my modification makes it so you end up with N points over the 2pi azimuthal angle and N/2 over the pi elevation angle, so in a sense it looks/is more uniform (by which in this case I mean what the spacing of the points ends up looking like). It's probably equivalent to creating a meshgrid of $\phi$ and $\theta$ directly.
